src/gil.rs

```diff
@@ -3,12 +3,33 @@
 //! Interaction with python's global interpreter lock
 
 use crate::{ffi, internal_tricks::Unsendable, PyAny, Python};
+use std::cell::Cell;
 use std::ptr::NonNull;
 use std::{any, sync};
 
 static START: sync::Once = sync::Once::new();
 static START_PYO3: sync::Once = sync::Once::new();
 
+thread_local! {
+    /// This is a internal counter in pyo3 monitoring whether this thread has the GIL.
+    ///
+    /// It will be incremented whenever a GIL-holding RAII struct is created, and decremented
+    /// whenever they are dropped.
+    ///
+    /// As a result, if this thread has the GIL, GIL_COUNT is greater than zero.
+    static GIL_COUNT: Cell<u32> = Cell::new(0);
+}
+
+/// Check whether the GIL is acquired.
+///
+/// Note: This uses pyo3's internal count rather than PyGILState_Check for two reasons:
+///  1) for performance
+///  2) PyGILState_Check always returns 1 if the sub-interpreter APIs have ever been called,
+///     which could lead to incorrect conclusions that the GIL is held.
+pub(crate) fn gil_is_acquired() -> bool {
```

Comment:
Why is this function `pub(crate)`?

Reply:
I wanted to be conservative about what we add to our public API. 

I thought other modules inside `pyo3` might need it, which is why I chose `pub(crate) fn` rather than just `fn`.

What would you prefer this to be?

Reply:
> I thought other modules inside pyo3 might need it, which is why I chose pub(crate) fn rather than just fn.
What would you prefer this to be?

I'm sorry but now I cannot judge which is better since I can't imagine other usages of this function.
I was just worried about the case you forgot to remove `pub(crate)`. If not, I want to let you decide.

Reply:
Ok for now I will just go for `fn`. 👍